src/runtime/importhook.cs

```diff
@@ -115,187 +97,154 @@ internal static void SaveRuntimeData(RuntimeDataStorage storage)
+                        Runtime.XDecref(pyNs);
+                    }
                 }
-            }
-            Runtime.XIncref(py_clr_module);
-            return NewReference.DangerousFromPointer(py_clr_module);
-        }
-
-        /// <summary>
-        /// The actual import hook that ties Python to the managed world.
-        /// </summary>
-        public static IntPtr __import__(IntPtr self, IntPtr argsRaw, IntPtr kw)
-        {
-            var args = new BorrowedReference(argsRaw);
 
-            // Replacement for the builtin __import__. The original import
-            // hook is saved as this.py_import. This version handles CLR
-            // import and defers to the normal builtin for everything else.
-
-            var num_args = Runtime.PyTuple_Size(args);
-            if (num_args < 1)
-            {
-                return Exceptions.RaiseTypeError("__import__() takes at least 1 argument (0 given)");
+                if (Runtime.PyDict_SetItemString(root.dict, availableNsKey, newset.DangerousGetAddress()) != 0)
+                {
+                    throw PythonException.ThrowLastAsClrException();
+                }
             }
-
-            BorrowedReference py_mod_name = Runtime.PyTuple_GetItem(args, 0);
-            if (py_mod_name.IsNull ||
-                !Runtime.IsStringType(py_mod_name))
+            finally
             {
-                return Exceptions.RaiseTypeError("string expected");
+                newset.Dispose();
             }
 
-            // Check whether the import is of the form 'from x import y'.
-            // This determines whether we return the head or tail module.
+        }
 
-            BorrowedReference fromList = default;
-            var fromlist = false;
-            if (num_args >= 4)
-            {
-                fromList = Runtime.PyTuple_GetItem(args, 3);
-                if (fromList != null &&
-                    Runtime.PyObject_IsTrue(fromList) == 1)
-                {
-                    fromlist = true;
-                }
-            }
+        /// <summary>
+        /// Removes the set of available namespaces from the clr module.
+        /// </summary>
+        static void TeardownNameSpaceTracking()
+        {
+            // If the C# runtime isn't loaded, then there are no namespaces available
+            Runtime.PyDict_SetItemString(root.dict, availableNsKey, Runtime.PyNone);
+        }
 
-            string mod_name = Runtime.GetManagedString(py_mod_name);
-            // Check these BEFORE the built-in import runs; may as well
-            // do the Incref()ed return here, since we've already found
-            // the module.
-            if (mod_name == "clr")
+        public static void AddNamespace(string name)
+        {
+            using (Py.GIL())
             {
-                NewReference clr_module = GetCLRModule(fromList);
-                if (!clr_module.IsNull())
+                var pyNs = Runtime.PyString_FromString(name);
```

Comment:
NIT: `using var pyNs = NewReference.DangerousFromPoiner(Runtime.PyString_FromString(name))` would remove both `try ... finally ...` and a cast for `PySet_Add`